common/lib/client/auth.ts

```diff
@@ -1,112 +1,135 @@
 import Logger from '../util/logger';
 import Platform from 'platform';
-import Utils from '../util/utils';
+import * as Utils from '../util/utils';
 import Http from 'platform-http';
 import Multicaster from '../util/multicaster';
-import BufferUtils from 'platform-bufferutils';
+import * as BufferUtils from 'platform-bufferutils';
 import ErrorInfo from '../types/errorinfo';
 import Base64 from 'platform-base64';
 import HmacSHA256 from 'crypto-js/build/hmac-sha256';
 import { stringify as stringifyBase64 } from 'crypto-js/build/enc-base64';
-
-var Auth = (function() {
-	var MAX_TOKEN_LENGTH = Math.pow(2, 17);
-	function noop() {}
-	function random() { return ('000000' + Math.floor(Math.random() * 1E16)).slice(-16); }
-	function normaliseAuthcallbackError(err) {
-		/* A client auth callback may give errors in any number of formats; normalise to an errorinfo */
-		if(!Utils.isErrorInfo(err)) {
-			return new ErrorInfo(Utils.inspectError(err), err.code || 40170, err.statusCode || 401);
-		}
-		/* network errors will not have an inherent error code */
-		if(!err.code) {
-			if(err.statusCode === 403) {
-				err.code = 40300;
-			} else {
-				err.code = 40170;
-				/* normalise statusCode to 401 per RSA4e */
-				err.statusCode = 401;
-			}
-		}
-		return err;
-	}
-
-	var hmac, toBase64;
-	if(Platform.createHmac) {
-		toBase64 = function(str) { return (Buffer.from(str, 'ascii')).toString('base64'); };
-		hmac = function(text, key) {
-			var inst = Platform.createHmac('SHA256', key);
-			inst.update(text);
-			return inst.digest('base64');
-		};
-	} else {
-		toBase64 = Base64.encode;
-		hmac = function(text, key) {
-			return stringifyBase64(HmacSHA256(text, key));
-		};
+import { createHmac } from 'crypto';
+import { ErrnoException, RequestCallback, RequestParams } from '../../types/http';
+import * as API from '../../../ably';
+import { StandardCallback } from '../../types/utils';
+
+// TODO: replace these with the real types once these classes are in TypeScript
```

Comment:
Lots of TODOs like this that look like they were intended to be dealt with before doing the final typescript release?

Reply:
mb, addressed them all now